README.md

```diff
@@ -18,3 +18,397 @@ We are looking for more hands to join in - or review the task list so that we ca
+nodes_array=($nodes)
+
+head_node=${nodes_array[0]}
+ip=$(srun --nodes=1 --ntasks=1 -w "$head_node" hostname --ip-address)
+
+# Normalize to IPv4 if multiple IPs
+if [[ "$ip" == *" "* ]]; then
+  IFS=' ' read -ra ADDR <<< "$ip"
+  if [[ ${#ADDR[0]} -gt 16 ]]; then ip=${ADDR[1]}; else ip=${ADDR[0]}; fi
+  echo "Multiple IPs detected; using IPv4 $ip"
+fi
+
+port=6379
+ip_head="$ip:$port"
+export ip_head
+export RAY_ADDRESS="$ip_head"
+echo "Head node: $head_node | IP Head: $ip_head"
+
+# ===== Start HEAD (daemon step that watches SENTINEL) =====
+echo "STARTING HEAD at ${head_node}"
+srun --export=ALL --nodes=1 --ntasks=1 -w "$head_node" \
+  bash --noprofile --norc -lc '
+    set -euo pipefail
+    SENTINEL="'"$SENTINEL"'"
+    # On TERM/INT from Slurm, stop Ray gracefully then exit 0.
+    cleanup_node() { ray stop -v --grace-period 60 || true; exit 0; }
+    trap cleanup_node TERM INT
+
+    # Start Ray head and watch for shutdown
+    ray start --head -v --node-ip-address="'"$ip"'" --port="'"$port"'" \
+      --redis-password="'"$redis_password"'" --block &
+    ray_pid=$!
+
+    while :; do
+      # If sentinel exists, stop Ray and exit 0
+      [[ -f "$SENTINEL" ]] && cleanup_node
+      # If Ray died for any reason, just exit 0 (step completes cleanly)
+      if ! kill -0 "$ray_pid" 2>/dev/null; then exit 0; fi
+      sleep 1
+    done
+  ' &
+
+# ===== Wait for head ready =====
+echo "Waiting for Ray head to be ready..."
+for i in {1..40}; do
+  if ray status --address "$ip_head" &>/dev/null; then
+    echo "Ray head is ready."
+    break
+  fi
+  if [[ $i -eq 40 ]]; then
+    echo "Ray head failed to become ready in time." >&2
+    exit 1
+  fi
+  sleep 2
+done
+
+# ===== Start WORKERS (each watches the same SENTINEL) =====
+worker_num=$((SLURM_JOB_NUM_NODES - 1))
+if (( worker_num > 0 )); then
+  for ((i = 1; i <= worker_num; i++)); do
+    node_i=${nodes_array[$i]}
+    echo "STARTING WORKER $i at $node_i"
+    srun --export=ALL --nodes=1 --ntasks=1 -w "$node_i" \
+      bash --noprofile --norc -lc '
+        set -euo pipefail
+        SENTINEL="'"$SENTINEL"'"
+        cleanup_node() { ray stop -v --grace-period 60 || true; exit 0; }
+        trap cleanup_node TERM INT
+
+        ray start -v --address="'"$ip_head"'" --redis-password="'"$redis_password"'" --block &
+        ray_pid=$!
+
+        while :; do
+          [[ -f "$SENTINEL" ]] && cleanup_node
+          if ! kill -0 "$ray_pid" 2>/dev/null; then exit 0; fi
+          sleep 1
+        done
+      ' &
+    # small gap avoids thundering herd on startup
+    sleep 1
+  done
+fi
+
+echo "All Ray nodes started. Giving workers a moment to register..."
+sleep 5
+
+# ===== Run your workload =====
+echo "Running user's command: ${COMMAND_PLACEHOLDER}"
+${COMMAND_PLACEHOLDER}
+exit_code=$?
+
+echo "User command finished with exit code $exit_code."
+echo "Shutting down Ray cluster (creating sentinel)..."
+# (trap will create the sentinel and wait for steps)
+exit $exit_code
+```
```

Comment:
I'm confused about what I'm supposed to do with this file. Am I supposed to customize it somehow? Do I just need to create and run it? What exactly does it do? Seems like it's creating a Ray cluster out of the slurm container nodes we set up earlier with docker compose, but it'd be nice to not have to read through the script to figure that out. Also, could this just be a file in the repo, instead of embedded in the README?

Reply:
Did you run it on your side? Did things work for you?

Reply:
moved to files already - the readme later on needs further cleanup though